posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+---
+layout: post
+title: An IO monad for cats
+
+meta:
+  nav: blog
+  author: djspiewak
+  pygments: true
+
+tut:
+  scala: 2.12.1
+  binaryScala: "2.12"
+  dependencies:
+    - org.scala-lang:scala-library:2.12.1
+    - org.typelevel::cats-effect:0.1-0848c9b
+---
+
+
+Haskell is a pure language.  Every Haskell expression is *referentially transparent*, meaning that you can substitute that expression with its evaluated result without changing the program.  Or, put into code:
+
+```haskell
+f e e
+-- is equivalent to
+let x = e in f x x
+```
+
+And this is true for *all* expressions `e`, and all functions `f`.  These could be complex expressions which describe ways of manipulating network channels or window buffers, or something trivial like a numeric literal.  You can *always* substitute the expression with its value.
+
+This is not true in Scala, simply because Scala allows *unrestricted* side-effects.  Unlike Haskell, Scala puts no limitations on where and when we can use things like mutable state (`var`s) or evaluated external effects like `println` or `launchTheMissiles`.  Since there are no restrictions on where and when we can do evil, the Scala equivalent to the above just doesn't work:
+
+```scala
+f(e, e)
+// isn't really equivalent to!
+val x = e
+f(x, x)
+```
+
+The reason it isn't equivalent comes from the different sorts of expressions that we could find in `e`.  For example, what if `e` is `println("hi!")`.  If we make that substitution, our snippet looks like the following:
+
+```scala
+f(println("hi"), println("hi"))
+// isn't really equivalent to!
+val x = println("hi")
+f(x, x)
+```
+
+Clearly these are not the same two programs.  The first prints `"hi"` twice, while the second only prints it once.  This is a violation of referential transparency, and it's why we sometimes say that Scala is an *impure* language.
```

Comment:
Bah, what does Wikipedia know?

> I should be more precise about what IO.apply does: captures expressions which contain side-effects.

I think you get this point across pretty well, but yes, the louder the better. This is the operation `Future` doesn't have.